release.sh

```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
+
+# release.sh will hopefully allow us to publish all of the necessary crates in
+# this repo in the right order, along with a few checks and balances to try to
+# avoid mistakes. It is assumed that only one person will be releasing all
+# crates at the same time.
+#
+# For each crate, it will:
+# 1. Run `cargo publish --dry-run` for that crate
+# 2. List all files in the package with `cargo package --list`
+# 3. Prompt the user as to whether to publish or not
+# 4. Publish the package with `cargo publish` (no dry run)
+#
+# It has a default set of crates it will publish, which can be overridden by
+# way of command line arguments:
+#
+#   # Release all packages, prompting for each package as to whether to publish
+#   ./release.sh
+#
+#   # Just release the proto and tendermint crates, but nothing else
+#   ./release.sh proto tendermint
+#
+# Once it publishes a crate, it will create a file at
+# /tmp/tendermint-rs-release/${TODAY}/${CRATE}, where ${TODAY} is today's date
+# and ${CRATE} is the name of the crate that was successfully published.
+#
+# Prior to publishing a crate, it checks whether this file is present before
+# attempting to publish it. If it's present, it will ask if you really want to
+# publish it again. Of course, this is pretty dumb, and doesn't cater for
+# instances where multiple people could publish the crates on the same day, and
+# instances where someone reboots their machine or wipes their /tmp folder
+# between runs.
+
+set -e
+
+ARGS="$@"
+TODAY=$(date +%Y-%m-%d)
+TMP_RELEASE_FOLDER="/tmp/tendermint-rs-release/${TODAY}"
```

Comment:
alternative:

```suggestion
TMP_RELEASE_FOLDER="$(mktemp -d -t tendermint-rs-release)"
```

Reply:
I'm also a fan of using `mktemp` for temporary folders, but in this case it takes away the feature of checking if there was a release already today. (A subsequent run would create a new temporary folder, never being able to check for older files.)

I'm more inclined to check on `crates.io` instead if a release has happened before, possibly using an API call. Which would also require a check in the source code for the version number of the crate. That can be done with `cargo metadata`.

I feel at this point this script is better suited for private use: it definitely shortens the release time but you need to be aware of our release process and a lot of assumptions have to be correct.

Probably I would like to see a script that does some analysis&print first. For example, use case:
- If I run the script without any parameters, it prints out all crates that are supposed to be published, their version numbers in the source code and a check if these version numbers are already published on crates.io. (Or possibly the latest version that's published on crates.io.)
- If I run the script with `--dry-run` or similar, it does the same and also runs the package list and publish with dry-run.
- If I run the script with `--publish`, it does the above and then publishes it.
- If I add package names on the command line, it filters down to those only.

I understand that adding all this might overcomplicate the script, so I think the most use we'd get from the first feature: gather and print as much info as possible. Even if we don't implement the actual publish or dry-run commands, it gives a much better handle on where we are in the publishing process.

With all that said, I'm happy to just use this script and continue improving this one instead. Just my two cents.